File: apps/mobile/lib/features/products/models/product.dart
// Product domain models for the Mazhavil Costumes mobile app.
//
// Mirrors the Next.js Product domain types including costumes-specific
// fields (material, metal_purity, metal_color, weight_grams) and
// rental-specific fields (condition, min/max rental days, inventory counts).

class BranchInventory {
  final String id;
  final String productId;
  final String branchId;
  final int stockCount;
  final String? branchName;
  final String createdAt;
  final String updatedAt;

  BranchInventory({
    required this.id,
    required this.productId,
    required this.branchId,
    required this.stockCount,
    this.branchName,
    required this.createdAt,
    required this.updatedAt,
  });

  factory BranchInventory.fromJson(Map<String, dynamic> json) {
    // Extract branch name from nested 'branches' object if present
    String? branchName = json['branch_name'] as String?;
    if (branchName == null && json['branches'] is Map) {
      branchName = (json['branches'] as Map<String, dynamic>)['name'] as String?;
    }

    return BranchInventory(
      id: json['id'] as String? ?? '',
      productId: json['product_id'] as String? ?? '',
      branchId: json['branch_id'] as String? ?? '',
      stockCount: json['stock_count'] as int? ?? json['quantity'] as int? ?? 0,
      branchName: branchName,
      createdAt: json['created_at'] as String? ?? '',
      updatedAt: json['updated_at'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'product_id': productId,
      'branch_id': branchId,
      'stock_count': stockCount,
      'branch_name': branchName,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}

class ProductImage {
  final String url;
  final String altText;
  final bool isPrimary;
  final int sortOrder;

  ProductImage({
    required this.url,
    this.altText = '',
    this.isPrimary = false,
    this.sortOrder = 0,
  });

  factory ProductImage.fromJson(dynamic json) {
    if (json is String) {
      return ProductImage(url: json);
    }
    return ProductImage(
      url: json['url'] ?? '',
      altText: json['alt_text'] ?? json['alt'] ?? '',
      isPrimary: json['is_primary'] ?? false,
      sortOrder: json['sort_order'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'url': url,
      'alt_text': altText,
      'is_primary': isPrimary,
      'sort_order': sortOrder,
    };
  }
}

class ProductVariant {
  final String id;
  final String name;
  final String value;
  final double additionalPrice;
  final bool inStock;

  ProductVariant({
    required this.id,
    required this.name,
    required this.value,
    this.additionalPrice = 0,
    this.inStock = true,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> json) {
    return ProductVariant(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      value: json['value'] as String? ?? '',
      additionalPrice: (json['additional_price'] as num?)?.toDouble() ?? 0,
      inStock: json['in_stock'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'value': value,
      'additional_price': additionalPrice,
      'in_stock': inStock,
    };
  }
}

class Product {
  final String id;
  final String storeId;
  final String? categoryId;
  final String? subcategoryId;
  final String? subvariantId;
  final String? branchId;
  final String name;
  final String slug;
  final String? description;
  final String? sku;
  final String? barcode;
  final double pricePerDay;
  final double securityDeposit;

  // Inventory
  final int quantity;
  final int availableQuantity;
  final int reservedQuantity;
  final int maintenanceQuantity;
  final int totalQuantity;

  // Costumes-specific
  final String? material;
  final String? metalPurity;
  final String? metalColor;
  final double? weightGrams;

  // Rental-specific
  final String? condition;
  final String? sanitizationStatus;
  final int? minRentalDays;
  final int? maxRentalDays;

  // Media & variants
  final List<ProductImage> images;
  final List<ProductVariant> sizes;
  final List<ProductVariant> colors;

  // Flags
  final bool isActive;
  final bool isFeatured;
  final bool trackInventory;
  final int lowStockThreshold;

  // Timestamps
  final String createdAt;
  final String? updatedAt;

  // Branch inventory
  final List<BranchInventory> branchInventory;

  Product({
    required this.id,
    required this.storeId,
    this.categoryId,
    this.subcategoryId,
    this.subvariantId,
    this.branchId,
    required this.name,
    required this.slug,
    this.description,
    this.sku,
    this.barcode,
    required this.pricePerDay,
    this.securityDeposit = 0,
    required this.quantity,
    required this.availableQuantity,
    this.reservedQuantity = 0,
    this.maintenanceQuantity = 0,
    this.totalQuantity = 0,
    this.material,
    this.metalPurity,
    this.metalColor,
    this.weightGrams,
    this.condition,
    this.sanitizationStatus,
    this.minRentalDays,
    this.maxRentalDays,
    this.images = const [],
    this.sizes = const [],
    this.colors = const [],
    this.isActive = true,
    this.isFeatured = false,
    this.trackInventory = true,
    this.lowStockThreshold = 10,
    required this.createdAt,
    this.updatedAt,
    this.branchInventory = const [],
  });

  /// Quick helper — primary image URL or null.
  String? get primaryImageUrl {
    if (images.isEmpty) return null;
    final primary = images.where((i) => i.isPrimary);
    return primary.isNotEmpty ? primary.first.url : images.first.url;
  }

  /// Stock status label for UI badges.
  String get stockStatusLabel {
    if (availableQuantity <= 0) return 'Out of Stock';
    if (availableQuantity <= lowStockThreshold) return 'Low Stock';
    return 'In Stock';
  }

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      storeId: json['store_id'] as String? ?? '',
      categoryId: json['category_id'] as String?,
      subcategoryId: json['subcategory_id'] as String?,
      subvariantId: json['subvariant_id'] as String?,
      branchId: json['branch_id'] as String?,
      name: json['name'] as String,
      slug: json['slug'] as String? ?? '',
      description: json['description'] as String?,
      sku: json['sku'] as String?,
      barcode: json['barcode'] as String?,
      pricePerDay: (json['price_per_day'] as num?)?.toDouble() ?? 0.0,
      securityDeposit: (json['security_deposit'] as num?)?.toDouble() ?? 0.0,
      quantity: json['quantity'] as int? ?? 0,
      availableQuantity: json['available_quantity'] as int? ?? 0,
      reservedQuantity: json['reserved_quantity'] as int? ?? 0,
      maintenanceQuantity: json['maintenance_quantity'] as int? ?? 0,
      totalQuantity: json['total_quantity'] as int? ?? json['quantity'] as int? ?? 0,
      material: json['material'] as String?,
      metalPurity: json['metal_purity'] as String?,
      metalColor: json['metal_color'] as String?,
      weightGrams: (json['weight_grams'] as num?)?.toDouble(),
      condition: json['condition'] as String?,
      sanitizationStatus: json['sanitization_status'] as String?,
      minRentalDays: json['min_rental_days'] as int?,
      maxRentalDays: json['max_rental_days'] as int?,
      images: (json['images'] as List<dynamic>?)
              ?.map((e) => ProductImage.fromJson(e))
              .toList() ??
          [],
      sizes: (json['sizes'] as List<dynamic>?)
              ?.map((e) => ProductVariant.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      colors: (json['colors'] as List<dynamic>?)
              ?.map((e) => ProductVariant.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      isActive: json['is_active'] as bool? ?? true,
      isFeatured: json['is_featured'] as bool? ?? false,
      trackInventory: json['track_inventory'] as bool? ?? true,
      lowStockThreshold: json['low_stock_threshold'] as int? ?? 10,
      createdAt: json['created_at'] as String? ?? '',
      updatedAt: json['updated_at'] as String?,
      branchInventory: _parseBranchInventory(json),
    );
  }

  /// Parse branch inventory from either 'branch_inventory' or 'product_inventory' key.
  static List<BranchInventory> _parseBranchInventory(Map<String, dynamic> json) {
    // Try 'branch_inventory' first (mobile API), then 'product_inventory' (Supabase join)
    final raw = json['branch_inventory'] ?? json['product_inventory'];
    if (raw is List) {
      return raw.map((e) => BranchInventory.fromJson(e as Map<String, dynamic>)).toList();
    }
    return [];
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'store_id': storeId,
      'category_id': categoryId,
      'subcategory_id': subcategoryId,
      'subvariant_id': subvariantId,
      'branch_id': branchId,
      'name': name,
      'slug': slug,
      'description': description,
      'sku': sku,
      'barcode': barcode,
      'price_per_day': pricePerDay,
      'security_deposit': securityDeposit,
      'quantity': quantity,
      'available_quantity': availableQuantity,
      'reserved_quantity': reservedQuantity,
      'maintenance_quantity': maintenanceQuantity,
      'total_quantity': totalQuantity,
      'material': material,
      'metal_purity': metalPurity,
      'metal_color': metalColor,
      'weight_grams': weightGrams,
      'condition': condition,
      'sanitization_status': sanitizationStatus,
      'min_rental_days': minRentalDays,
      'max_rental_days': maxRentalDays,
      'images': images.map((e) => e.toJson()).toList(),
      'sizes': sizes.map((e) => e.toJson()).toList(),
      'colors': colors.map((e) => e.toJson()).toList(),
      'is_active': isActive,
      'is_featured': isFeatured,
      'track_inventory': trackInventory,
      'low_stock_threshold': lowStockThreshold,
      'created_at': createdAt,
      'updated_at': updatedAt,
      'branch_inventory': branchInventory.map((e) => e.toJson()).toList(),
    };
  }
}

File: apps/admin/domain/types/product.ts
/**
 * Product Domain Types
 *
 * Type definitions for the product domain following Domain-Driven Design principles.
 * These types represent the core business entities and value objects.
 *
 * @module domain/types/product
 */

// Core Product Entity
export interface Product {
  readonly id: string;
  readonly store_id: string;
  readonly category_id: string | null;
  readonly subcategory_id: string | null;
  readonly subvariant_id: string | null;
  readonly branch_id: string | null;
  name: string;
  slug: string;
  description: string | null;
  sku: string | null;
  barcode: string | null;
  price_per_day: number;
  purchase_price: number;
  quantity: number;
  available_quantity: number;
  images: ProductImage[];
  sizes: ProductVariant[];
  colors: ProductVariant[];
  is_active: boolean;
  is_featured: boolean;
  track_inventory: boolean;
  low_stock_threshold: number;
  created_at: string;
  updated_at?: string;
  // Audit fields
  readonly created_by: string | null;
  readonly created_at_branch_id: string | null;
  readonly updated_by: string | null;
  readonly updated_at_branch_id: string | null;
}

// Product Image Value Object
export interface ProductImage {
  url: string;
  alt_text?: string;
  is_primary: boolean;
  sort_order: number;
}

// Product Variant Value Object
export interface ProductVariant {
  id: string;
  name: string;
  value: string;
  additional_price?: number;
  in_stock: boolean;
}

// Product Status Enum
export enum ProductStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  DRAFT = 'draft',
  ARCHIVED = 'archived',
}

// Product Inventory Status
export enum InventoryStatus {
  IN_STOCK = 'in_stock',
  LOW_STOCK = 'low_stock',
  OUT_OF_STOCK = 'out_of_stock',
  DISCONTINUED = 'discontinued',
}

// Product Create DTO (Data Transfer Object)
export interface CreateProductDTO {
  name: string;
  slug: string;
  sku?: string;
  barcode?: string;
  category_id?: string;
  subcategory_id?: string;
  subvariant_id?: string;
  branch_id?: string | 'all';
  store_id?: string;
  description?: string;
  price_per_day: number;
  purchase_price?: number;
  quantity?: number;
  available_quantity?: number;
  images?: Omit<ProductImage, 'id'>[];
  sizes?: Omit<ProductVariant, 'id'>[];
  colors?: Omit<ProductVariant, 'id'>[];
  is_active?: boolean;
  is_featured?: boolean;
  track_inventory?: boolean;
  low_stock_threshold?: number;
  branch_inventory?: Array<{ branch_id: string; quantity: number; id?: string }>;
}

// Product Update DTO
export interface UpdateProductDTO {
  name?: string;
  slug?: string;
  sku?: string;
  barcode?: string;
  category_id?: string | null;
  subcategory_id?: string | null;
  subvariant_id?: string | null;
  description?: string;
  price_per_day?: number;
  purchase_price?: number;
  quantity?: number;
  available_quantity?: number;
  images?: Omit<ProductImage, 'id'>[];
  sizes?: Omit<ProductVariant, 'id'>[];
  colors?: Omit<ProductVariant, 'id'>[];
  is_active?: boolean;
  is_featured?: boolean;
  track_inventory?: boolean;
  low_stock_threshold?: number;
  branch_inventory?: Array<{ branch_id: string; quantity: number; id?: string }>;
  removed_inventory_ids?: string[];
}

// Product Search Parameters
export interface ProductSearchParams {
  query?: string;
  category_id?: string;
  store_id?: string;
  branch_id?: string;
  status?: ProductStatus;
  is_featured?: boolean;
  min_price?: number;
  max_price?: number;
  in_stock?: boolean;
  sort_by?: 'name' | 'price' | 'created_at' | 'stock';
  sort_order?: 'asc' | 'desc';
  page?: number;
  limit?: number;
}

// Product Search Result
export interface ProductSearchResult {
  products: Product[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
  total_stock?: number;
}

// Product with Related Data
export interface ProductWithRelations extends Product {
  category?: {
    id: string;
    name: string;
    slug: string;
  };
  store?: {
    id: string;
    name: string;
  };
}

// Product Inventory Info
export interface ProductInventory {
  product_id: string;
  total_quantity: number;
  available_quantity: number;
  reserved_quantity: number;
  sold_quantity: number;
  status: InventoryStatus;
  low_stock_threshold: number;
  last_updated: string;
}

// Product Pricing Info
export interface ProductPricing {
  product_id: string;
  price_per_day: number;
  purchase_price: number;
  currency: string;
  discount_percentage?: number;
  valid_from?: string;
  valid_to?: string;
}

// Product Analytics
export interface ProductAnalytics {
  product_id: string;
  views: number;
  orders: number;
  revenue: number;
  average_rating: number;
  total_reviews: number;
  conversion_rate: number;
  period: {
    start: string;
    end: string;
  };
}

// Product Bulk Operations
export interface BulkProductOperation {
  product_ids: string[];
  operation: 'activate' | 'deactivate' | 'delete' | 'feature' | 'unfeature' | 'update_price';
  data?: Record<string, any>;
}

// Product Bulk Operation Result
export interface BulkOperationResult {
  successful: string[];
  failed: Array<{
    product_id: string;
    error: string;
  }>;
  total_processed: number;
  total_successful: number;
  total_failed: number;
}

// Product Import/Export
export interface ProductImportData {
  products: CreateProductDTO[];
  total_count: number;
  valid_count: number;
  invalid_count: number;
  errors: Array<{
    row: number;
    field: string;
    message: string;
  }>;
}

export interface ProductExportData {
  products: Product[];
  export_format: 'csv' | 'xlsx' | 'json';
  filters: ProductSearchParams;
  exported_at: string;
}

// Product Validation Errors
export interface ProductValidationError {
  field: string;
  message: string;
  code: string;
}

// Product Validation Result
export interface ProductValidationResult {
  is_valid: boolean;
  errors: ProductValidationError[];
  warnings: ProductValidationError[];
}

// Domain Events
export interface ProductDomainEvent {
  type: 'product_created' | 'product_updated' | 'product_deleted' | 'product_activated' | 'product_deactivated';
  product_id: string;
  data: Record<string, any>;
  timestamp: string;
  user_id?: string;
}

// Product Aggregate Root (for complex business logic)
export interface ProductAggregate {
  product: Product;
  inventory: ProductInventory;
  pricing: ProductPricing;
  analytics: ProductAnalytics;
  
  // Business logic methods
  canBeDeleted(): boolean;
  isInStock(): boolean;
  isLowStock(): boolean;
  calculateTotalPrice(days: number): number;
  applyDiscount(percentage: number): void;
  reserveStock(quantity: number): boolean;
  releaseStock(quantity: number): void;
}

// Type Guards
export const isValidProduct = (obj: any): obj is Product => {
  return obj && 
         typeof obj.id === 'string' &&
         typeof obj.store_id === 'string' &&
         typeof obj.name === 'string' &&
         typeof obj.slug === 'string' &&
         typeof obj.price_per_day === 'number' &&
         typeof obj.quantity === 'number' &&
         typeof obj.available_quantity === 'number';
};

export const isActiveProduct = (product: Product): boolean => {
  return product.is_active && product.available_quantity > 0;
};

export const isLowStockProduct = (product: Product): boolean => {
  return product.track_inventory && 
         product.available_quantity <= product.low_stock_threshold;
};

export const isOutOfStockProduct = (product: Product): boolean => {
  return product.track_inventory && product.available_quantity === 0;
};
